File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    max: int = 0
    totalSum: int = 0
    MOD: int = 1_000_000_007
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        # Get all sum's value...
        self.totalSum = self.getSum(root)
        # Find maximized product...
        self.dfs(root)
        return self.max % self.MOD
        
    def getSum(self,root:Optional[TreeNode]) -> int:
        if(root==None): return 0
        return root.val+(self.getSum(root.left))+(self.getSum(root.right))

    def dfs(self,root:Optional[TreeNode]) -> int:
        if(root==None): return 0
        leftSum: int = self.dfs(root.left)
        rightSum: int = self.dfs(root.right)
        self.max = max(self.max,(self.totalSum-leftSum)*leftSum)
        self.max = max(self.max,(self.totalSum-rightSum)*rightSum)
        return leftSum+rightSum+root.val

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private long max;
    private static final int MOD = 1_000_000_007;
    long totalSum;
    public int maxProduct(TreeNode root) {
        // let's sum all the values in a tree...
        totalSum = getSum(root);
        // Find maximum value...
        dfs(root);
        return (int)(max % MOD);
    }
    private long getSum(TreeNode root){
        if(root==null) return 0;
        return (long)root.val+getSum(root.left)+getSum(root.right);
    }
    private int dfs(TreeNode root){
        if(root==null) return 0;
        int leftSum = dfs(root.left);
        int rightSum = dfs(root.right);
        max = Math.max(max,(totalSum-leftSum)*leftSum);
        max = Math.max(max,(totalSum-rightSum)*rightSum);
        return root.val+leftSum+rightSum;
    }
}